janus_collector/src/lib.rs

```diff
@@ -0,0 +1,866 @@
+//! PPM protocol collector
+
+use backoff::{future::retry, ExponentialBackoff};
+use derivative::Derivative;
+use janus_core::{
+    hpke::{self, associated_data_for_aggregate_share, HpkeApplicationInfo, HpkePrivateKey},
+    message::{CollectReq, CollectResp, HpkeConfig, Interval, Role, TaskId},
+    retries::{http_request_exponential_backoff, retry_http_request},
+};
+use prio::{
+    codec::{Decode, Encode},
+    vdaf,
+};
+use reqwest::{
+    header::{HeaderValue, ToStrError, CONTENT_TYPE, LOCATION},
+    StatusCode,
+};
+use std::time::Duration as StdDuration;
+use url::Url;
+
+/// HTTP header where auth tokens are provided in messages to the leader.
+pub const DAP_AUTH_HEADER: &str = "DAP-Auth-Token";
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error("HTTP client error: {0}")]
+    HttpClient(#[from] reqwest::Error),
+    #[error("HTTP response status {0}")]
+    Http(StatusCode),
+    #[error("URL parse: {0}")]
+    Url(#[from] url::ParseError),
+    #[error("Missing Location header in See Other response")]
+    MissingLocationHeader,
+    #[error("Invalid bytes in header")]
+    InvalidHeader(#[from] ToStrError),
+    #[error("Wrong Content-Type header: {0:?}")]
+    BadContentType(Option<HeaderValue>),
+    #[error("Codec error: {0}")]
+    Codec(#[from] prio::codec::CodecError),
+    #[error("Aggregate share decoding error")]
+    AggregateShareDecode,
+    #[error("Expected two aggregate shares, got {0}")]
+    AggregateShareCount(usize),
+    #[error("VDAF error: {0}")]
+    Vdaf(#[from] prio::vdaf::VdafError),
+    #[error("HPKE error: {0}")]
+    Hpke(#[from] janus_core::hpke::Error),
+    #[error("Timed out waiting for collection to finish")]
+    CollectPollTimeout,
+}
+
+static COLLECTOR_USER_AGENT: &str = concat!(
+    env!("CARGO_PKG_NAME"),
+    "/",
+    env!("CARGO_PKG_VERSION"),
+    "/",
+    "collector"
+);
+
+/// Authentication configuration for communication with the leader aggregator.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[non_exhaustive]
+pub enum Authentication {
+    /// Beearer token authentication, via the `DAP-Auth-Token` header.
+    BearerToken(#[derivative(Debug = "ignore")] String),
```

Comment:
I'm not sure it's correct to call this a bearer token since we're not doing `Authorization: Bearer <token>` but instead our bespoke `DAP-Auth-Token: <token>`. In any case, it seems like this should use `janus_server::task::AuthenticationToken`, moving that struct into `janus_core` as needed.

Reply:
+1 to a different name, `DAP-Auth-Token` is (in the generic sense) a bearer token but the potential for confusion is high.

We could be potentially overly-literal and call this variant literally `DAPAuthToken`; I guess the upside is that it has a high chance of not conflicting with other auth mechanisms introduced later.